src/hotspot/share/gc/shenandoah/shenandoahGenerationSizer.cpp

```diff
@@ -0,0 +1,209 @@
+  size_t min_young_regions = (heap_region_count * ShenandoahMinYoungPercentage) / 100;
+  return MAX2(min_young_regions, (size_t) 1U);
+}
+
+size_t ShenandoahGenerationSizer::calculate_max_young_regions(size_t heap_region_count) {
+  size_t max_young_regions = (heap_region_count * ShenandoahMaxYoungPercentage) / 100;
+  return MAX2(max_young_regions, (size_t) 1U);
+}
+
+void ShenandoahGenerationSizer::recalculate_min_max_young_length(size_t heap_region_count) {
+  assert(heap_region_count > 0, "Heap must be initialized");
+
+  switch (_sizer_kind) {
+    case SizerDefaults:
+      _min_desired_young_regions = calculate_min_young_regions(heap_region_count);
+      _max_desired_young_regions = calculate_max_young_regions(heap_region_count);
+      break;
+    case SizerNewSizeOnly:
+      _max_desired_young_regions = calculate_max_young_regions(heap_region_count);
+      _max_desired_young_regions = MAX2(_min_desired_young_regions, _max_desired_young_regions);
+      break;
+    case SizerMaxNewSizeOnly:
+      _min_desired_young_regions = calculate_min_young_regions(heap_region_count);
+      _min_desired_young_regions = MIN2(_min_desired_young_regions, _max_desired_young_regions);
+      break;
+    case SizerMaxAndNewSize:
+      // Do nothing. Values set on the command line, don't update them at runtime.
+      break;
+    case SizerNewRatio:
+      _min_desired_young_regions = MAX2(uint(heap_region_count / (NewRatio + 1)), 1U);
+      _max_desired_young_regions = _min_desired_young_regions;
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+
+  assert(_min_desired_young_regions <= _max_desired_young_regions, "Invalid min/max young gen size values");
+}
+
+void ShenandoahGenerationSizer::heap_size_changed(size_t heap_size) {
+  recalculate_min_max_young_length(heap_size / ShenandoahHeapRegion::region_size_bytes());
+}
+
+bool ShenandoahGenerationSizer::transfer_regions(ShenandoahGeneration* src, ShenandoahGeneration* dst, size_t regions) const {
+  const size_t bytes_to_transfer = regions * ShenandoahHeapRegion::region_size_bytes();
+
+  if (src->free_unaffiliated_regions() < regions) {
+    // Source does not have enough free regions for this transfer. The caller should have
+    // already capped the transfer based on available unaffiliated regions.
+    return false;
+  }
+
+  if (dst->max_capacity() + bytes_to_transfer > max_size_for(dst)) {
+    // This transfer would cause the destination generation to grow above its configured maximum size.
+    return false;
+  }
+
+  if (src->max_capacity() - bytes_to_transfer < min_size_for(src)) {
+    // This transfer would cause the source generation to shrink below its configured minimum size.
+    return false;
+  }
+
+  src->decrease_capacity(bytes_to_transfer);
+  dst->increase_capacity(bytes_to_transfer);
+  const size_t new_size = dst->max_capacity();
+  log_info(gc)("Transfer " SIZE_FORMAT " region(s) from %s to %s, yielding increased size: " PROPERFMT,
```

Comment:
This should not be `log_info(gc)`. In fact, please look at other places where you print logging. Generally, `log_info(gc)` should provide _only_ the high-level GC info: which phases were running. Everything else should go under `log_debug(gc)`.

Reply:
https://bugs.openjdk.org/browse/JDK-8344263